### PR TITLE
#996 Crypto symbol length

### PR DESCRIPTION
**Pull request by viliwonka**
#### Description
Extended DecomposeCurrencyPair in Forex.cs code, so that it supports 3-4 chars per symbol/currency. Modified tests a little so they also support 4 chars. 

Also added a bit of crypto symbols and pairs (also with length of 4), so that DecomposeCurrencyPair gets tested by them - and those crypto - pairs will get added in future anyway.

#### Related Issue

https://github.com/QuantConnect/Lean/issues/996

#### Motivation and Context

Cryptocurrencies tickers have 3 or 4 chars per symbol. Which makes pair string consisting of 6-8 pairs. 
Unit tests in Lean fail because of this, since DecomposeCurrencyPair doesn't work on such strings.
Now currency pair strings can have length from 6-8.

#### Requires Documentation Change

Changed documentation above DecomposeCurrencyPair, also it's required to add cryptocurrency symbols into Currencies.CurrencySymbols for this to work - it will fail for pair strings longer than 6 chars.

#### How Has This Been Tested?

Testing environment: Windows 10, C#, no Python binaries installed.
Python tests were failing, almost all others are ok. Same tests pass as with freshly pulled Lean source code, so I assume I haven't broken anything. 

I have modified ForexTests.cs tests a little, so be sure to check that one.

#### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/ect)

#### Checklist:

- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [x] I have added tests to cover my changes. 
- [ ] All new and existing tests passed.
- [x] My branch follows the naming convention `bug-<issue#>-<description` or `feature-<issue#>-<description>`

#### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/ect)

#### Checklist:

- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [x] I have added tests to cover my changes. 
- [x] All new and existing tests passed.
- [ ] My branch follows the naming convention `bug-<issue#>-<description` or `feature-<issue#>-<description>`

### CODE REVIEW

**Comment by viliwonka** [comment]
My first PR.

All tests that I think have passed, but not all passed. 

But at same time, I pulled fresh QuantConnect code, did unit tests on it, and same tests fail, so I suppose I haven't broken anything.

**Comment by mchandschuh** [comment]
It seems travis is failing this. Here's the logs from travis:
https://api.travis-ci.org/v3/job/377363331/log.txt

**Comment by viliwonka** [comment]
Ah, last minute adding new pairs, but without having a symbols. My mistake. Will fix this asap.

File: Common/Currencies.cs [hunk 1]
```diff
@@ -303,7 +332,7 @@ public static class Currencies
         public static string GetCurrencySymbol(string currency)
         {
             string currencySymbol;
-            return CurrencySymbols.TryGetValue(currency, out currencySymbol) ? currencySymbol : "$";
+            return CurrencySymbols.TryGetValue(currency, out currencySymbol) ? currencySymbol : "?";
```
[the remark this answers is not in the document]
Comment (viliwonka) [reply Common/Currencies.cs]: Ok, changed back to $.

File: Common/Currencies.cs [hunk 2]
```diff
@@ -21,24 +21,46 @@ namespace QuantConnect
...
+    public static class Currencies {
         /// <summary>
         /// Gets the listing of crypto pairs used for currency conversion rates
         /// </summary>
         /// <remarks>
         /// It must include all currency pairs needed to resolve quote currencies in <see cref="Cash.EnsureCurrencyDataFeed"/>
         /// </remarks>
-        public static readonly IReadOnlyList<string> CryptoCurrencyPairs = new List<string>
-        {
+        public static readonly IReadOnlyList<string> CryptoCurrencyPairs = new List<string>() {
```
Comment (AlexCatarino) [inline Common/Currencies.cs]: Since there are other fixes to be made, you could undo this change here because of the style we follow.

Comment (viliwonka) [reply Common/Currencies.cs]: Missed that one, will check everything for code style mismatch again.

File: Common/Currencies.cs [hunk 3]
```diff
@@ -21,24 +21,47 @@ namespace QuantConnect
     /// <summary>
     /// Provides commonly used currency pairs and symbols
     /// </summary>
-    public static class Currencies
-    {
+    public static class Currencies {
```
Comment (AlexCatarino) [inline Common/Currencies.cs]: I didn't get this other style issue. Sorry. 
You may want to fix the last commit to include this: `git commit --fixup=65a86b1`

Comment (viliwonka) [reply Common/Currencies.cs]: Sharp eyes. No problem, .. will have to reconfigure my editor it seems.

Comment (mchandschuh) [reply Common/Currencies.cs]: love seeing the `git fu` grow :D

**Comment by mchandschuh** [comment]
This is looking good, perhaps we should run a test live algorithm with forex/crypto to ensure there are no regressions in the live system.

**Comment by mchandschuh** [comment]
Approved, will merge following successful live forex/crypto test in the QC cloud.

**Comment by AlexCatarino** [comment]
@viliwonka , I merged this PR and try to run an algorithm, but it failed in the [Cash ctor](https://github.com/QuantConnect/Lean/blob/master/Common/Securities/Cash.cs#L85).
Here is the algorithm I used so that you can reproduce the issue:
```csharp
namespace QuantConnect
{
    public class BasicTemplateAlgorithm : QCAlgorithm
    {
        public override void Initialize()
        {
            SetStartDate(2018, 1, 1);
            SetEndDate(2018, 4, 30);
            SetCash(100000);

            AddForex("EURUSD", Resolution.Minute);
	    AddCrypto("BTCUSD", Resolution.Minute);
	    AddCrypto("RHOCBTC", Resolution.Minute);
	    AddCrypto("SENTETH", Resolution.Minute);
        }

	public override void OnData(Slice data)
        {
            foreach(var symbol in data.Keys)
            {
                if (!Portfolio[symbol].Invested)
            	{
            	    SetHoldings(symbol, .1);
                    Debug($"Purchased {symbol}");
            	}
            }
        }
    }
}
```

**Comment by viliwonka** [comment]
Excuse me, completely missed this issue. Thanks for spotting it 👍 

Potential fix:

In Cash.cs, modify at line 83 the code:
```
    if (symbol == null || symbol.Length != 3)
    {
        throw new ArgumentException("Cash symbols must be exactly 3 characters.");
    }
```
to:

```
    if (symbol == null || symbol.Length < 3 || symbol.Length > 4)
    {
        throw new ArgumentException("Cash symbols must have atleast 3 characters and at most 4 characters.");
    }
```

Also pushed this modification to my fork:
https://github.com/viliwonka/Lean/commit/8b4b2ede552ec884a56e79acd9f95fc3940b0b80

Just didn't wanted to make another PR for such small fix.

**Comment by mchandschuh** [comment]
Please also run the entire regression suite of tests. I think this error would have been spotted via running the regression algorithms. Also, we may want to update `CashTests.cs` to cover these cases in unit tests so travis will catch them for us :)

**Comment by viliwonka** [comment]
@mchandschuh Ok, I updated CashTests.cs too and ran tests, but running the Alexs code, I saw the issues are deeper than it looks at first look. My code works fine, but it seems as if splitting of ticker is also done somewhere else.

For example, it splits "RHOCBTC" into "RHO" and "CBTC" somewhere between Cash.cs and "EnsureCurrencyDataFeed". Investigating further.

**Comment by StefanoRaggi** [comment]
@viliwonka Since we're increasing the symbol length limit, I wonder if we should at least increase it to 10 instead of 8 (6 for base currency and 4 for quote currency).

`CHEESE/USDT` may not be one of the best pairs to trade, but it is listed on a few minor exchanges. 😄

**Comment by viliwonka** [comment]
Ah, had feeling that 4 chars won't be enough. Maybe we should make max length limit a variable instead of hardcoding it everywhere.
